src/app/blog/blob-store-design-challenges/page.mdx

```diff
@@ -0,0 +1,215 @@
+
+This is not the end of the world, but it does mean that we need to keep the data for a hash in some weird indeterminate state before we have enough data to decide if it will be stored in memory or on disk.
+
+# Blob lifecycle
+
+There are two fundamentally different ways how data can be added to a blob store. 
+
+## Adding local files by name
+
+If we add data locally, e.g. from a file, we have the data but don't know the hash. We have to compute the complete outboard, the root hash, and then atomically move the file into the store under the root hash. Depending on the size of the file, data and outboard will end up in the file system or in the database. If there was partial data there before, we can just replace it with the new complete data and outboard. Once the data is stored under the hash, neither the data nor the outboard will be changed.
+
+## Syncing remote blobs by hash
+
+If we sync data from a remote node, we do know the hash but don't have the data. In case the blob is small, we can request and atomically write the data, so we have a similar situation as above. But as soon as the data is larger than a chunk group size (16 KiB), we will have a situation where the data has to be written incrementally. This is much more complex than adding local files. We now have to keep track of which chunks of the blob we have locally, and which chunks of the blob we can *prove* to have locally (not the same thing if you have a chunk group size > 1).
+
+## Using the data
+
+As soon as data is added, we can serve it over the network or use it locally. We don't have to wait for completion for this. We can start sharing a large blob even before we are finished downloading it.
+
+When a range request comes in that contains ranges that are not yet present locally, we will notice that the local data is incomplete and stop answering the request - outgoing data is always validated.
+
+Alternatively we can rely on a bitfield to make sure that all requested data is present locally.
+
+## Blob deletion
+
+On creation, blobs are tagged with a temporary tag that prevents them from being deleted for as long as the process lives. They can then be tagged with a persistent tag that prevents them from being deleted even after a restart. And last but not least, large groups of blobs can be protected from deletion in bulk by putting a sequence of hashes into a blob and tagging that blob as a hash sequence.
```

Comment:
```suggestion
On creation, blobs are tagged with a temporary tag that prevents them from being deleted for as long as the process lives. They can then be tagged with a persistent tag that extends their persistence beyond restarts. And last but not least, large groups of blobs can be protected from deletion in bulk by putting a sequence of hashes into a blob and tagging that blob as a hash sequence.
```

Reply:
"prevents them from being deleted even after a restart" explains what is meant by persistence. So I like the previous version better.